File: algorithms/0532.k-diff-pairs-in-an-array/solution.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def find_pairs(nums, k)
  res = 0

  hash = Hash.new(0)
  nums.each { |n| hash[n] += 1 }

  hash.each do |key, value|
    if hash.has_key?(key + k)
      if (k == 0 && value > 1) || k > 0
        res += 1
      end
    end
  end

  res
end

File: algorithms/0532.k-diff-pairs-in-an-array/solution.go
func findPairs(nums []int, k int) int {
	m := map[int]int{}
	for _, n := range nums {
		if _, ok := m[n]; ok {
			m[n] += 1
		} else {
			m[n] = 1
		}
	}

	res := 0
	for key, value := range m {
		if _, ok := m[key+k]; ok {
			if (k == 0 && value > 1) || k > 0 {
				res++
			}
		}
	}

	return res
}
